### PR TITLE
eliminate warnings for latest openssl and gcc

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
Signed-off-by: SchrodingerZhu <i@zhuyi.fan>

### CODE REVIEW

**Comment by SchrodingerZhu** [comment]
we dont have CI for this?

**Comment by SchrodingerZhu** [comment]
Changes follow upstream

File: Foundation/src/MD4Engine.cpp [hunk 1]
```diff
@@ -130,8 +130,8 @@ const DigestEngine::Digest& MD4Engine::digest()
 	/* Store state in digestArray */
 	unsigned char digestArray[16];
 	encode(digestArray, _context.state, 16);
-	_digest.clear();
-	_digest.insert(_digest.begin(), digestArray, digestArray + sizeof(digestArray));
+    _digest.resize(sizeof(digestArray));
```
Comment (fuzhe1989) [inline Foundation/src/MD4Engine.cpp]: how about `_digest.assign(begin(digestArray), end(digestArray))`?

File: Foundation/src/MD4Engine.cpp [hunk 2]
```diff
@@ -130,8 +130,8 @@ const DigestEngine::Digest& MD4Engine::digest()
...
+    std::copy(digestArray, digestArray + sizeof(digestArray), _digest.begin());
```
Comment (fuzhe1989) [inline Foundation/src/MD4Engine.cpp]: ditto

File: Foundation/src/MD5Engine.cpp [hunk 3]
```diff
@@ -130,8 +130,8 @@ const DigestEngine::Digest& MD5Engine::digest()
 	/* Store state in digestArray */
 	unsigned char digestArray[16];
 	encode(digestArray, _context.state, 16);
-	_digest.clear();
-	_digest.insert(_digest.begin(), digestArray, digestArray + sizeof(digestArray));
+	_digest.resize(sizeof(digestArray));
+	std::copy(digestArray, digestArray + sizeof(digestArray), _digest.begin());
```
Comment (fuzhe1989) [inline Foundation/src/MD5Engine.cpp]: ditto

File: Foundation/src/SHA1Engine.cpp [hunk 4]
```diff
@@ -142,8 +142,8 @@ const DigestEngine::Digest& SHA1Engine::digest()
 	unsigned char hash[DIGEST_SIZE];
 	for (count = 0; count < DIGEST_SIZE; count++)
 		hash[count] = (BYTE) ((_context.digest[count>>2]) >> (8*(3-(count & 0x3)))) & 0xff;
-	_digest.clear();
-	_digest.insert(_digest.begin(), hash, hash + DIGEST_SIZE);
+    _digest.resize(sizeof(hash));
+    std::copy(hash, hash + sizeof(hash), _digest.begin());
```
Comment (fuzhe1989) [inline Foundation/src/SHA1Engine.cpp]: ditto

File: Foundation/src/MD4Engine.cpp (hunk 1 above)
Comment (SchrodingerZhu) [reply Foundation/src/MD4Engine.cpp]: great！

File: Foundation/src/MD4Engine.cpp (hunk 2 above)
Comment (SchrodingerZhu) [reply Foundation/src/MD4Engine.cpp]: ok

**Comment by fuzhe1989** [comment]
LGTM

**Review by fuzhe1989 — APPROVED** [review]

**Review by hanfei1991 — APPROVED** [review]

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Pull request merged**